File: 백준/Silver/11055. 가장 큰 증가하는 부분 수열/가장 큰 증가하는 부분 수열.java
import java.io.*;
import java.util.*;

public class Main {

    static int n;
    /**
     * 주의할 점!!!
     * 우리가 구하려고 하는거는 n에서의 증가하는 부분 수열의 최대 합이 아니고, 이 수열 전체에서의 최대 합
     * i번째를 포함했다면, arr[i]보다 작은 arr[k] 여러 개를 찾은 뒤, 그 중 최대를 찾는다.
     * O(n^2)
     */
    static int[] dp = new int[1002];
    static int[] arr = new int[1002];
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
            int localMax = 0;
            for (int j = 0; j < i; j++) {
                if (arr[j] < arr[i]) localMax = Math.max(localMax, dp[j]);
            }
            dp[i] = localMax + arr[i];
        }
        System.out.print(Arrays.stream(dp, 0, n).parallel().max().getAsInt());
    }
}

File: 백준/Silver/11055. 가장 큰 증가하는 부분 수열/가장 큰 증가하는 부분 수열.kt
/**
 * d[i]: 
 * - i가 선택되었을 때, 최대 합
 * d[i] = d[k] + a[i]
 * - k는 i보다 작지만, 가장 d[k]가 큰 원소
 */

import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    val arr = StringTokenizer(readLine()).let { st -> IntArray(n) { st.nextToken().toInt() } }
    val d = IntArray(n)

    for (i in 0 until n) {
        d[i] = arr[i]
        for (j in i - 1 downTo 0) {
            if (arr[j] < arr[i]) {
                d[i] = maxOf(d[i], d[j] + arr[i])
            }
        }
    }
    print(d.max())
}